### DISCUSSION TITLE
Any way to get a result that "extend" a trip response by a distance on the same road?

### DISCUSSION THREAD

**Discussion by Dunge** [valhalla/valhalla#4466] [Q&A]
Hello all, a bit new to this so sorry if this is a stupid question.

I'm using the `trace_route` api request to match a list of semi-imprecise gps coordinate into a most possibly traveled route and it works great. I only use for the returned shape object that return the travel path, not the maneuvers.

What I would like is to be able to pass a parameter that is a distance before the first point and after the last one, let's say "add 500 meters" onto the same road and direction to extend that path (of course, assuming the road extend that long). Is that something I could do one way or another with Valhalla?

I have two ideas:
- Manually calculate a vector based on the direction of the two last points in my original coordinate list and project a final points further away. That would lead to a very imprecise result in case the road curves, it would embark on another one.
- Use the `Isodistances` service and pass the returned first and last points. Filter everything except the edge with a `way_id` that matches the one of these points. Then call `trace_route` a second time with these new points to get the final shape. I guess it could work, but it seems a bit overkill to calculate all those possible paths when I only requires one, and would require 3 API calls instead of just one.

Is there a much easier method I'm just unaware of?

Could using the Overpass API help me achieve that better instead of Valhalla?

**Comment by nilsnolde** [comment]
Interesting problem:) can I ask the motivation?

I guess there’s a few ways to achieve that, you’re already mentioning 2. Instead of isochrones/-distances you could use /locate on the end points, which should give you all the info of the roads at origin/destination you’d need. That won’t work if you’d like to continue onto another connected road. In that case you could just call the /expansion service with 500 isodistance, which will return the road geometries of all connected roads within 500 m and you could walk the connected edges in the other direction. That service returns the edge_ids and previous edge_ids (internal graph IDs), not the OSM IDs but that should be enough.

**Reply by nilsnolde** [reply]
But yeah it’s not possible to do that in the trace_route action. It’s such a rare case it wouldn’t be worth it implement either IMO. At least there’s alternatives:)

**Reply by Dunge** [reply]
Thank you.

Motivation : I work for a workzone equipment manufacturer business. We have connected devices with gps deployed along workzones and I use their positions to resolve a full path of the workzones. So my coordinate waypoints list aren't a travel path, but a list of device positions. But of course, equipments aren't always deployed at the limits of the construction site, so I need to offers a method for our managers to say "that workzone real end is X meters after the last variable message sign or radar".

Anyway, I'm not used to osm data yet, but calling the /locate api seems to return an "edge id" and the percent along it, but I don't quite see how it could helps me get the polyline of the path up to x meters before or after the requested point on that edge. Also, I believe edges are just small segments and a single road can be composed of many edges? So I would need to walk the graph of ids until I reach the wanted distance? I really don't understand how this would work.

**Reply by nilsnolde** [reply]
> I believe edges are just small segments and a single road can be composed of many edges?

That's right. Edges are the features between intersections, while a road, as we think about it, can stretch many intersections.

> So I would need to walk the graph of ids until I reach the wanted distance? 

Jep, you'd need to walk it exactly. That's what I was alluding to with the /expansion endpoint: https://valhalla.github.io/valhalla/api/expansion/api-reference/. It returns a list of edge_ids and previous edge_ids along with the geometry, which you can use to do exactly that, walk the graph, starting from your end edge.

**Reply by Dunge** [reply]
Been a while since I asked the question, but just wanted to add a few observations after actually trying it.

The expansion service works and returns all the edges that can be reached in that distance. But the returned shape is not limited to the distance, it shows the whole edge. So if the edge is 2km long, and you only ask for 200m, it returns the whole 2km anyway. Which kinda defeat my purpose because I can't use it as a precise distance delimitation.

I checked a whole other of isodistance mapping service apis, and the vast majority returns a polygon of the zone instead of the road network itself, so in that aspect Valhalla is still kinda better.
